lib/SILOptimizer/PassManager/PassPipeline.cpp

```diff
@@ -493,6 +493,8 @@ static void addHighLevelFunctionPipeline(SILPassPipelinePlan &P) {
   addFunctionPasses(P, OptimizationLevelKind::HighLevel);
 
   addHighLevelLoopOptPasses(P);
+  
+  P.addStringOptimization();
```

Comment:
It doesn't need to be before Serialization, it just has to be before the mid-level pipeline: the reason is that, even if StringOptimization is at the begin of the mid-level function pipeline, it can be that the whole mid-level function pipeline already has run for the global initializer (which means that the string init is inlined). And then the optimization cannot analyze the initializer anymore.

As we don't have a function-pipeline between serialization and the midlevel function pipeline, I put it into the high-level function pipeline.
@meg-gupta Feel free to change that in https://github.com/apple/swift/commit/7c9e8b8492fc3666b75e22791891ec802689dd9a. The pipeline just needs to make sure that all string operations are not inlined (including in the initializers) when StringOptimization runs.